File: frontend/src/api/chainsaw.ts
import api from './client'

export interface ChainsawScan {
  id:          string
  file_id:     string
  case_id:     string
  status:      'pending' | 'scanning' | 'ready' | 'error'
  alert_count: number | null
  error_msg:   string | null
  scanned_at:  string | null
  created_at:  string
}

export interface ChainsawAlert {
  id:           string
  scan_id:      string
  file_id:      string
  case_id:      string
  rule_name:    string
  level:        string | null   // critical | high | medium | low | informational
  sigma_status: string | null
  group_name:   string | null
  tags:         string | null
  authors:      string | null
  timestamp:    string | null
  event_id:     number | null
  channel:      string | null
  computer:     string | null
  provider:     string | null
  event_data:   Record<string, string> | null
  added_to_timeline: boolean
}

/** Alert enriched with source filename — stored in the pinned selection. */
export interface PinnedChainsawAlert extends ChainsawAlert {
  _filename: string
}

export interface AlertsPage {
  total:     number
  page:      number
  page_size: number
  pages:     number
  items:     ChainsawAlert[]
}

export interface ChainsawSelection {
  alert_ids: string[]
  sent_ids:  string[]
}

export interface AlertFilters {
  file_id?:   string
  levels?:    string   // comma-separated
  search?:    string
  sort_dir?:  'asc' | 'desc'
  page?:      number
  page_size?: number
}

export interface RuleInfo {
  filename: string
  title: string
  group: string
  level: string
  status: string
  description: string
  authors: string
  kind: string
  path: string
}

export interface SigmaStatus {
  installed: boolean
  rule_count: number
  sigma_dir: string
}

export const chainsawRulesApi = {
  listBuiltin: (): Promise<RuleInfo[]> =>
    api.get('/chainsaw/rules/builtin').then(r => r.data),

  listCustom: (): Promise<RuleInfo[]> =>
    api.get('/chainsaw/rules/custom').then(r => r.data),

  uploadCustom: (files: File[]): Promise<{ saved: string[] }> => {
    const fd = new FormData()
    files.forEach(f => fd.append('files', f))
    return api.post('/chainsaw/rules/custom/upload', fd).then(r => r.data)
  },

  deleteCustom: (filename: string): Promise<{ deleted: string }> =>
    api.delete(`/chainsaw/rules/custom/${encodeURIComponent(filename)}`).then(r => r.data),

  sigmaStatus: (): Promise<SigmaStatus> =>
    api.get('/chainsaw/rules/sigma/status').then(r => r.data),

  sigmaDownload: (): Promise<{ status: string; message: string }> =>
    api.post('/chainsaw/rules/sigma/download').then(r => r.data),
}

export const chainsawApi = {
  startScan: (caseId: string, fileId: string): Promise<ChainsawScan> =>
    api.post<ChainsawScan>(`/chainsaw/${caseId}/files/${fileId}/scan`).then(r => r.data),

  listScans: (caseId: string): Promise<ChainsawScan[]> =>
    api.get<ChainsawScan[]>(`/chainsaw/${caseId}/scans`).then(r => r.data),

  deleteScan: (caseId: string, scanId: string): Promise<void> =>
    api.delete(`/chainsaw/${caseId}/scans/${scanId}`).then(() => undefined),

  listAlerts: (caseId: string, filters: AlertFilters = {}): Promise<AlertsPage> =>
    api.get<AlertsPage>(`/chainsaw/${caseId}/alerts`, { params: filters }).then(r => r.data),

  sendToTimeline: (caseId: string, alertId: string): Promise<ChainsawAlert> =>
    api.post<ChainsawAlert>(`/chainsaw/${caseId}/alerts/${alertId}/timeline`).then(r => r.data),

  getSelection: (caseId: string): Promise<ChainsawSelection> =>
    api.get<ChainsawSelection>(`/chainsaw/${caseId}/selection`).then(r => r.data),

  saveSelection: (caseId: string, alert_ids: string[], sent_ids: string[]): Promise<ChainsawSelection> =>
    api.put<ChainsawSelection>(`/chainsaw/${caseId}/selection`, { alert_ids, sent_ids }).then(r => r.data),
}

File: backend/app/models/chainsaw.py
import uuid
from datetime import UTC, datetime

from sqlalchemy import JSON, Boolean, Column, DateTime, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import relationship

from app.database import Base


class ChainsawScan(Base):
    """One Chainsaw scan run against one EVTX file."""
    __tablename__ = "chainsaw_scans"

    id          = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    file_id     = Column(String, ForeignKey("evtx_files.id",  ondelete="CASCADE"), nullable=False, index=True)
    case_id     = Column(String, ForeignKey("cases.id",        ondelete="CASCADE"), nullable=False, index=True)
    status      = Column(String(16), nullable=False, default="pending")  # pending|scanning|ready|error
    alert_count = Column(Integer, nullable=True)
    error_msg   = Column(Text,    nullable=True)
    scanned_at  = Column(DateTime, nullable=True)
    created_at  = Column(DateTime, default=lambda: datetime.now(UTC))

    alerts = relationship(
        "ChainsawAlert", back_populates="scan",
        cascade="all, delete-orphan", lazy="dynamic",
    )


class ChainsawAlert(Base):
    """One Sigma rule match returned by a Chainsaw scan."""
    __tablename__ = "chainsaw_alerts"

    id           = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    scan_id      = Column(String, ForeignKey("chainsaw_scans.id", ondelete="CASCADE"), nullable=False, index=True)
    file_id      = Column(String, ForeignKey("evtx_files.id",    ondelete="CASCADE"), nullable=False)
    case_id      = Column(String, ForeignKey("cases.id",          ondelete="CASCADE"), nullable=False, index=True)

    # Sigma rule metadata
    rule_name    = Column(String(512), nullable=False, index=True)
    level        = Column(String(32),  nullable=True, index=True)   # critical/high/medium/low/informational
    sigma_status = Column(String(32),  nullable=True)               # stable/test/experimental
    group_name   = Column(String(255), nullable=True)
    tags         = Column(Text, nullable=True)                      # comma-separated
    authors      = Column(Text, nullable=True)

    # Event fields
    timestamp  = Column(DateTime, nullable=True, index=True)
    event_id   = Column(Integer,  nullable=True)
    channel    = Column(String(255), nullable=True)
    computer   = Column(String(255), nullable=True)
    provider   = Column(String(512), nullable=True)
    event_data = Column(JSON, nullable=True)   # all EventData/UserData fields

    added_to_timeline = Column(Boolean, default=False, nullable=False)

    scan = relationship("ChainsawScan", back_populates="alerts")

    __table_args__ = (
        Index("ix_chainsaw_alerts_case_ts",    "case_id", "timestamp"),
        Index("ix_chainsaw_alerts_case_level", "case_id", "level"),
        Index("ix_chainsaw_alerts_scan_level", "scan_id", "level"),
    )


class ChainsawCaseSelection(Base):
    """Persists the analyst's pinned Chainsaw alert selection per case (one row per case)."""
    __tablename__ = "chainsaw_case_selections"

    case_id    = Column(String, ForeignKey("cases.id", ondelete="CASCADE"), primary_key=True)
    alert_ids  = Column(JSON, nullable=False, default=list)   # list[str] of alert UUIDs
    sent_ids   = Column(JSON, nullable=False, default=list)   # alert UUIDs already pushed to timeline
    updated_at = Column(DateTime,
                        default=lambda:  datetime.now(UTC),
                        onupdate=lambda: datetime.now(UTC),
                        nullable=False)
